### PR TITLE
Change staking contract reward accrual mechanism

### PR DESCRIPTION
**Pull request by chuddster**

### CODE REVIEW

File: contracts/RibbonStakingRewards.sol [hunk 1]
```diff
@@ -27,7 +27,7 @@ contract StakingRewards is
     IERC20 public stakingToken;
     uint256 public periodFinish = 0;
     uint256 public rewardRate = 0;
-    uint256 public rewardsDuration = 7 days;
+    uint256 public rewardsDuration = 30 days;
```
Comment (chuddster) [inline contracts/RibbonStakingRewards.sol:30]: I initially thought we would use setRewardsDuration(duration) during the setup steps, but hardcoding is easier.

File: contracts/RibbonStakingRewards.sol [hunk 2]
```diff
@@ -27,7 +27,7 @@ contract StakingRewards is
...
     uint256 public lastUpdateTime;
     uint256 public rewardPerTokenStored;
```
Comment (chuddster) [inline contracts/RibbonStakingRewards.sol:33]: also, you will notice that in notifyRewardAmount, periodFinish is [defined](https://github.com/ribbon-finance/token/blob/dbe7357bca37d985647b9574b1548aae1bea362f/contracts/RibbonStakingRewards.sol#L206) based on current block timestamp. This is fine as long as we set notifyRewardAmount close to startEmissionAnyway. 

However, we can set it to periodFinish = startEmission.add(rewardDuration) for precision. but this means we cannot call notifyRewardAmount later on, so we need to be sure we will not be extending the LM program (extending entails calling notifyRewardAmount with new reward). this has been largely irrelevant until discussion of potential extension of program came up.

File: contracts/RibbonStakingRewards.sol [hunk 3]
```diff
@@ -157,7 +157,7 @@ contract StakingRewards is
     }
 
     function getReward() public override nonReentrant updateReward(msg.sender) {
-        uint256 reward = rewards[msg.sender];
+        uint256 reward = block.timestamp >= periodFinish ? rewards[msg.sender] : 0;
```
Comment (chuddster) [inline contracts/RibbonStakingRewards.sol:160]: quick one liner for locking solution. "locks" for entire month. if you withdraw beforehand you lose all rewards, but rewards are calculated based on how many weeks you were staking for.

Comment (kenchangh) [reply contracts/RibbonStakingRewards.sol:160]: Just to clarify, this doesn't actually forfeit the user's rewards right? What this change does is to ensure when `getReward` is called, 0 RBN is returned. This is desirable.

One change I recommend we do, we should add `rewards[msg.sender] = 0` when the user calls `withdraw`. This basically forfeits the user's rewards when the user withdraws

```solidity
function withdraw(uint256 amount)
    public
    override
    nonReentrant
    updateReward(msg.sender)
{
    rewards[msg.sender] = 0;
    require(amount > 0, "Cannot withdraw 0");
    _totalSupply = _totalSupply.sub(amount);
    _balances[msg.sender] = _balances[msg.sender].sub(amount);
    stakingToken.safeTransfer(msg.sender, amount);
    emit Withdrawn(msg.sender, amount);
}
```

Comment (chuddster) [reply contracts/RibbonStakingRewards.sol:160]: yes, you are right. technically they can go back and forth, depositing friday before reward and withdrawing right after. they cannot get the rewards, but they can get them much later and at the amount as if they stayed the whole time

**Review by kenchangh — DISMISSED** [review]

**Review by kenchangh — APPROVED** [review]

**Pull request merged**